infra/lambda/datasets-endpoint/function.py

```diff
@@ -0,0 +1,482 @@
+    db_resp = DYNAMODB.scan(
+        TableName="datasets",
+        ExpressionAttributeNames={
+            "#id": "pk",
+            "#type": "sk",
+        },
+        ExpressionAttributeValues={
+            ":id": {"S": "DATASET#"},
+            ":type": {"S": "TYPE#"},
+        },
+        FilterExpression="begins_with(#id, :id) and begins_with(#type, :type)",
+        Select="ALL_ATTRIBUTES",
+    )
+
+    resp_body = []
+    for i in db_resp["Items"]:
+        item = {}
+        item["id"] = list(i["pk"].values())[0].split("#")[1]
+        item["type"] = list(i["sk"].values())[0].split("#")[1]
+
+        for a in DS_ATTRIBUTES + DS_ATTRIBUTES_EXT:
+            item[a] = list(i[a].values())[0]
+
+        resp_body.append(item)
+
+    return {"statusCode": 200, "body": resp_body}
+
+
+def patch_method(payload):
+    """PATCH: Update Dataset."""
+
+    BODY_SCHEMA = {
+        "type": "object",
+        "properties": {
+            "id": {"type": "string"},
+            "type": {
+                "type": "string",
+                "enum": ["IMAGE", "RASTER"],
+            },
+            "title": {"type": "string"},
+            "owning_group": {"type": "string"},
+        },
+        "required": [
+            "id",
+            "type",
+        ],
+        "minProperties": 3,
+    }
+
+    # request body validation
+    try:
+        req_body = payload["body"]
+        validate(req_body, BODY_SCHEMA)
+    except ValidationError as err:
+        return {"statusCode": 400, "body": {"message": f"Bad Request: {err.message}."}}
+
+    # get PKs
+    pk = {}
+    pk["id"] = payload["body"]["id"]
+    pk["type"] = payload["body"]["type"]
+
+    # get attributes
+    attr = {}
+    for a in DS_ATTRIBUTES:
+        if a in payload["body"].keys():
+            attr[a] = payload["body"][a]
+
+    # make sure that requested type/title doesn't already exist in DB
+    db_resp = DYNAMODB.query(
+        TableName="datasets",
+        IndexName="datasets_title",
+        ExpressionAttributeNames={
+            "#type": "sk",
+            "#title": "title",
+        },
+        ExpressionAttributeValues={
+            ":type": {"S": f"TYPE#{pk['type']}"},
+            ":title": {"S": f"{attr['title']}"},
+        },
+        KeyConditionExpression="#title = :title AND #type = :type",
+        Select="COUNT",
+        ConsistentRead=False,
+    )
+    if int(db_resp["Count"]) > 0:
+        return {
+            "statusCode": 409,
+            "body": {
+                "message": f"Conflict: dataset '{attr['title']}' of type '{pk['type']}' already exists."
+            },
+        }
+
+    # update Dataset record in DB
+    expression_attribute_names = {}
+    for a in attr:
+        expression_attribute_names[f"#{a}"] = a
+
+    expression_attribute_values = {}
+    for a in attr:
+        expression_attribute_values[f":{a}"] = {"S": attr[a]}
+
+    update_expression = "SET "
+    update_expression += ", ".join([f"#{a} = :{a}" for a in attr])
```

Comment:
Looks like SQL injection but it would be much harder to succeed then in SQL I think. Right now I don't have any idea how to improve ti.

Reply:
It looks like the potential for injection is less devastating than regular SQL, but there are still [information leaks](https://medium.com/appsecengineer/dynamodb-injection-1db99c2454ac) to consider.

However, looking closer at how `attr` is generated we should be safe, because we only include attribute names which are in the dataset attributes. Combined with `expression_attribute_values` being a dictionary I think we're safe. :+1: